mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -582,24 +584,28 @@ private[spark] object RandomForest extends Logging {
 
         // Extract info for this node.  Create children if not leaf.
         val isLeaf =
-          (stats.gain <= 0) || (LearningNode.indexToLevel(nodeIndex) == metadata.maxDepth)
+          (stats.gain <= 0) || (node.level == metadata.maxDepth)
         node.isLeaf = isLeaf
         node.stats = stats
         logDebug("Node = " + node)
 
         if (!isLeaf) {
           node.split = Some(split)
-          val childIsLeaf = (LearningNode.indexToLevel(nodeIndex) + 1) == metadata.maxDepth
+          val childLevel = node.level + 1
+          val childIsLeaf = childLevel == metadata.maxDepth
           val leftChildIsLeaf = childIsLeaf || (stats.leftImpurity == 0.0)
           val rightChildIsLeaf = childIsLeaf || (stats.rightImpurity == 0.0)
-          node.leftChild = Some(LearningNode(LearningNode.leftChildIndex(nodeIndex),
-            leftChildIsLeaf, ImpurityStats.getEmptyImpurityStats(stats.leftImpurityCalculator)))
-          node.rightChild = Some(LearningNode(LearningNode.rightChildIndex(nodeIndex),
-            rightChildIsLeaf, ImpurityStats.getEmptyImpurityStats(stats.rightImpurityCalculator)))
+          val leftChildIndex = nodeIdAssigners(treeIndex).nextIndex()
+          val rightChildIndex = nodeIdAssigners(treeIndex).nextIndex()
+          node.leftChild = Some(LearningNode(leftChildIndex, leftChildIsLeaf,
+            ImpurityStats.getEmptyImpurityStats(stats.leftImpurityCalculator), childLevel))
+          node.rightChild = Some(LearningNode(rightChildIndex, rightChildIsLeaf,
+            ImpurityStats.getEmptyImpurityStats(stats.rightImpurityCalculator), childLevel))
```

Comment:
This function with its nested functions is a monster :)   Probably not worth attacking

Could encapsulate this adding-a-node code, but I think this is passable